src/Traversals/depthfirst.jl

```diff
@@ -95,14 +95,21 @@ end
     vcolor = zeros(UInt8, nv(g))
     verts = Vector{T}()
     state = TopoSortState(vcolor, verts, zero(T))
-    for v in vertices(g)
-        state.vcolor[v] != 0 && continue
-        state.vcolor[v] = 1
-        if !traverse_graph!(g, v, DepthFirst(), state)
-            throw(CycleError())
-        end
+    
+    sources = filter(x -> isempty(inneighbors(g, x)), vertices(g))
+   
+   
+       
+    if !traverse_graph!(g, sources , DepthFirst(), state)
+         throw(CycleError())
+    end
+      
```

Comment:
We can shortcut this: `traverse_graph(g, sources, DepthFirst(), state) || throw(CycleError())`